### ISSUE TITLE
Introduce @angularEntrypoint annotation to decide where to pull in ng_deps

### ISSUE THREAD

**Issue by yjbanov** [angular/angular#4865]
Currently we use poor man's pattern matching to decide which files to rewrite to swap out reflection with generated ng_deps. Instead, a developer should be able to select any file they consider an "entry point" - a test file, an app with a main method, something else - and tell us right in that file that we need to add ng_deps to imports.

## Proposal

Introduce `@angularEntrypoint` annotation that developer applies to a function or method:

```dart
library my_test;

@angularEntrypoint
main() {  // doesn't have to be `main`; any method/function will suffice
}
```

Our transformer would look for this annotation and rewrite the file by adding `ng_deps` imports, adding calls to `initReflector` functions, and swapping out `bootstrap` for `bootstrapStatic`. Even if there is no call to `bootstrap` we would still add `ng_deps`. This way people can bring their own custom bootstrap API (very popular in tests).

## Details

Let's say we have the following Dart file:

```dart
library has_entrypoint;

import 'a.dart';
import 'b.dart';

class App {
  @angularEntrypoint
  void start() {
    bootstrap(AComponent);
  }
}

@angularEntrypoint
main() {
  bootstrap(BComponent);
}
```

The transformer identifies the `@angularEntrypoint` annotation and rewrite this file to:

```dart
library has_entrypoint;

import 'a.dart';import 'a.ng_deps.dart' as i0;import 'b.ng_deps.dart' as i1;
import 'b.dart';

class App {
  @angularEntrypoint
  void start() {i0.initReflector();i1.initReflector();
    bootstrapStatic(AComponent);
  }
}

@angularEntrypoint
main() {i0.initReflector();i1.initReflector();
  bootstrapStatic(BComponent);
}
```

Notes:

-  `ng_deps` imports are injected in a way that does not alter line numbers for better debugging experience
- for the same reason calls to `initReflector` go on the same line as the opening curly of the function/method declaration
- both entry points are calling all `initReflector` even if they use only some of them (this is because otherwise we'd need to resolve the whole program to figure out which types are used where); it is up to developer to split their code into multiple libraries to get finer grained control over code size (in practice, there would rarely be production code with multiple entry points in the same file; this is mostly useful for tests)

**Comment by kegluneq** [comment]
I'm on board for the general approach.

Some minor nits:
- We will need to add one `initReflector` per `.ng_deps.dart` import we add unless we can determine where the `bootstrap`ed `Type` is coming from (we probably could by reading the summaries the transformer generates)
- We should try to avoid changing line numbers to keep source maps fresh, so these calls should happen on the same line as the opening bracket of the method labeled `@entrypoint`

**Comment by yjbanov** [comment]
@kegluneq good points; I updated the description

**Comment by yjbanov** [comment]
another edit: renamed to `@angularEntrypoint`

**Comment by mhevery** [comment]
I am fine with the proposal. My only comment is that it should be dart specific.

**Comment by tbosch** [comment]
Bumping priority to P0 as it prevents us from executing some tests in G3 as they are now using mirrors which makes the binary too big to run.

**Comment by tbosch** [comment]
Workaround for now: Add manual calls to `initReflector` of the `.ng_deps.dart` file.

**Comment by kegluneq** [comment]
re: my earlier comment, I was wrong about inserting multiple `initReflector` calls.

If the entrypoint is `a.dart`, we only need to call `a.ng_deps.dart#initReflector`, which will call its dependencies' `initReflector` methods.

**Comment by kegluneq** [comment]
Closed via 87449ab43c8547ea786713f5a2245b1ab2197afd
